File: STUDY/8. 정수론/1016. 제곱 ㄴㄴ 수/1016.cpp
#include <iostream>
#include <cmath>

using namespace std;

long long MIN, MAX, R;
int arr[int(1e6) + 1];

int main() {
	cin >> MIN >> MAX;
	R = 0;
	arr[int(1e6) + 1] = { 0, };
	for (long long i = 2; i < sqrt(MAX) + 1; i++) {
		long long P = i * i;
		long long S = MIN / P;
		if (MIN % P != 0) S++;
		for (long long j = S; j < (MAX / P) + 1; j++) {
			arr[j * P - MIN] = 1;
		}
	}
	for (int i = 0; i <= MAX - MIN; i++) {
		if (arr[i] == 0) R++;
	}
	cout << R << '\n';
	return 0;
}

File: STUDY/8. 정수론/1016. 제곱 ㄴㄴ 수/1016.py
MIN,MAX=map(int,input().split())
arr=[0]*(MAX-MIN+1)
for i in range(2,int(MAX**0.5)+1):
    P=i**2
    start=MIN//P
    if MIN%P: start+=1
    for j in range(start,MAX//P+1):
        arr[j*P-MIN]=1
print(arr.count(0))

# MIN,MAX=map(int,input().split())
# LIMIT=int(1e6)
# result=MAX-MIN+1
# arr=[0]*(LIMIT+1)
# Prime=[]
# for i in range(2,LIMIT+1):
#     if arr[i]: continue
#     if i**2<=int(1e6):
#         Prime+=[i**2]
#     j=2
#     while i*j<=int(1e6):
#         arr[i*j]=1
#         j+=1
# arr=[0]*(MAX-MIN+1)
# arr[0]=1
# for P in Prime:
#     j=0
#     while P*j<=MAX-MIN:
#         if not arr[P*j]:
#             arr[P*j]=1
#             result-=1
#         j+=1
# print(arr.count(0))